### PR TITLE
Improves Beam() code, Adds a new spacevine mutation+mob

### PR DESCRIPTION
**Pull request by RemieRichards**
- Adds a sanity check to ensure `origin` exists in Beam code
- Updates the `x_oldloc` vars so that Beams update as little as possible, previously they'd always check against where the beam was originally started, leading to a lot more `Reset()` and `Draw()` calls needlessly
- Makes `pixel_x/_y` calculations only happen if `DX/DY` are not 0, instead of calculating it and then discarding it if they were 0 (seriously, wtf)
- Makes Beams use a `W in X to Y step Z` construct for loop, the fastest loop type for this use case.
- Makes Beams anchored, because I noticed them being pushed by space wind...
- Makes Beams have `mouse_opacity = 0`, they're effects and shouldnt be clicked in most cases, and seeing them called "ebeam" is odd.
- Added a new spacevine mutation, a chance to spawn a flower bud that grows into a Venus Human Trap after 2 minutes
- Added Venus Human Traps, an enemy that grabs hold of it's enemies, weakening them, and pulling them towards itself (if it doesnt just run straight at you and bite your face off)

Sprites are... not perfect.
flowerbud

vht

### CODE REVIEW

**Comment by duncathan** [comment]
10/10 codersprites

**Comment by AnturK** [comment]
:+1:  I was pretty lazy in that rework

**Comment by Dorsisdwarf** [comment]
Does this mean we can make laser cannons beams?

**Comment by Iamgoofball** [comment]
when this gets merged I'm putting audrey 2s in the traitor uplink

**Comment by Razharas** [comment]
>New spacevine mutation

Yes, YES

**Comment by Niknakflak** [comment]
>more ways for bobdobbington to fuck the crew

Greeeaaattttt
but really :+1: and sprites remind me of plantera from terraria

**Comment by Shadowlight213** [comment]
Replace blob with sentient spacevines when?

**Comment by KorPhaeron** [comment]
Test merging

**Comment by RemieRichards** [comment]
@Niknakflak the bud/bulb flower actually *IS* based on Plantera.
@Shadowlight213 funnilly enough, if you read spacevine code they mention they're "very similar to (read: copy pasted from) 'biomass' code", biomass was a very VERY early version of blob, so these two things are actually cousins!

**Comment by Iamgoofball** [comment]
>Makes Beams have mouse_opacity = 0, they're effects and shouldnt be clicked in most cases, and seeing them called "ebeam" is odd.

what if we added chopping the vines with a hatchet to save people from becoming dentists

**Comment by RemieRichards** [comment]
@Iamgoofball if you read the code, the vine beams DONT share the ```mouse_opacity``` trait, because I gave them a name and a desc.

**Comment by Iamgoofball** [comment]
>reading code
>not just instant merging

**Comment by Kitsune-bi** [comment]
The real question is "can we feed the plant(s) a sadist competing for the same love interest?"

**Comment by scrdest** [comment]
As minor as it might be, 'human trap' sounds a bit clunky. Mantrap?

**Comment by Bluespace** [comment]
Sexist pig.

**Comment by Kitsune-bi** [comment]
Well for the name I don't see many good names from the "____ trap." You could name than something like "digesters" though that is more of a slang name like face huggers. Last thing I can think of is "trap plant".

It would also be interesting if the grabbing beams were also like feelers but having it so the plant can't see per say but the grabbers would feel around for moving stuff to eat but walking on vines that are a part of the plant tell it something is stepping on it and cause the grabbers to "feel" around that area for anything moving.

**Comment by scrdest** [comment]
@Bluespace Oink.

**Comment by Flip-Sandwich** [comment]
venus fly trap for nerds

**Comment by RemieRichards** [comment]
@duncathan 
1. DONT COMMENT ON COMMITS YOU SPAZ.
2. Not using a sqrt helper (we only have a proc for it atm, not an inline define) because speed.
3. 32 and 16 should be obviously, it's ```world.icon_size``` and ```world.icon_size*0.5```, but Gunbuddy wrote them straight out, I'm not changing it, it's fine.

File: code/datums/beam.dm [hunk 1]
```diff
@@ -26,53 +27,76 @@
...
 /datum/beam/Destroy()
 	Reset()
 	target = null
 	origin = null
 	return ..()
 
+
 /datum/beam/proc/Draw()
-	var/Angle=round(Get_Angle(origin,target))
+	var/Angle = round(Get_Angle(origin,target))
```
Comment (duncathan) [inline code/datums/beam.dm]: messing with comment chronology cuz I'm a dum dum

isn't there a helper for this?

File: code/datums/beam.dm [hunk 2]
```diff
@@ -26,53 +27,76 @@
...
-	var/DX=(32*target.x+target.pixel_x)-(32*origin.x+origin.pixel_x)
-	var/DY=(32*target.y+target.pixel_y)-(32*origin.y+origin.pixel_y)
-	var/N=0
-	var/length=round(sqrt((DX)**2+(DY)**2))
-	
-	for(N,N<length,N+=32)
-		var/obj/effect/ebeam/X= new beam_type(origin.loc)
-		X.owner=src
+	//Translation vector for origin and target
+	var/DX = (32*target.x+target.pixel_x)-(32*origin.x+origin.pixel_x)
```
Comment (duncathan) [inline code/datums/beam.dm]: what is 32 here? I don't read magic

File: code/datums/beam.dm [hunk 3]
```diff
@@ -26,53 +27,76 @@
...
-		if(DX==0) Pixel_x=0
-		if(DY==0) Pixel_y=0
+
+		//Calculate pixel offsets (If necessary)
+		var/Pixel_x
+		var/Pixel_y
+		if(DX == 0)
+			Pixel_x = 0
+		else
+			Pixel_x = round(sin(Angle)+32*sin(Angle)*(N+16)/32)
```
Comment (duncathan) [inline code/datums/beam.dm]: same with 16 here

File: code/datums/beam.dm (hunk 1 above)
Comment (RemieRichards) [inline code/datums/beam.dm]: did you comment on the right line here? ```Get_Angle()``` is the helper.

Comment (duncathan) [inline code/datums/beam.dm]: WRONG LINE I'M EVEN STUPIDER

File: code/datums/beam.dm [hunk 4]
```diff
@@ -26,53 +27,76 @@
...
-	var/length=round(sqrt((DX)**2+(DY)**2))
-	
-	for(N,N<length,N+=32)
-		var/obj/effect/ebeam/X= new beam_type(origin.loc)
-		X.owner=src
+	//Translation vector for origin and target
+	var/DX = (32*target.x+target.pixel_x)-(32*origin.x+origin.pixel_x)
+	var/DY = (32*target.y+target.pixel_y)-(32*origin.y+origin.pixel_y)
+	var/N = 0
+	var/length = round(sqrt((DX)**2+(DY)**2)) //hypotenuse of the triangle formed by target and origin's displacement
```
Comment (duncathan) [inline code/datums/beam.dm]: isn't there a helper for this?

Comment (RemieRichards) [inline code/datums/beam.dm]: see:

>2. Not using a sqrt helper (we only have a proc for it atm, not an inline define) or cheap_trig() (because it isn't cheap) because speed.

File: code/datums/beam.dm (hunk 2 above)
Comment (RemieRichards) [inline code/datums/beam.dm]: see:

>32 and 16 should be obviously, it's world.icon_size and world.icon_size*0.5, but Gunbuddy wrote them straight out, I'm not changing it, it's fine.

File: code/datums/beam.dm (hunk 3 above)
Comment (RemieRichards) [inline code/datums/beam.dm]: see: 

>32 and 16 should be obviously, it's world.icon_size and world.icon_size*0.5, but Gunbuddy wrote them straight out, I'm not changing it, it's fine.

File: code/datums/beam.dm (hunk 2 above)
Comment (duncathan) [inline code/datums/beam.dm]: >32 and 16 should be obviously, it's `world.icon_size` and `world.icon_size*0.5`

oh ok so this is X distance between them, measured in pixels? I was having trouble understanding that
maybe just a comment clarifying that would be helpful?

Comment (RemieRichards) [inline code/datums/beam.dm]: DY by DX is standard mathematical notation for DISTANCE ON X AXIS, DISTANCE ON Y AXIS, like high/secondary school level
It was obvious to me, since I'd seen this notation before, and the comment about the hypotenuse (below) already sort of clarifies.

File: code/datums/beam.dm (hunk 4 above)
Comment (duncathan) [inline code/datums/beam.dm]: we should make a macro for that cuz speed
this is cool then, tho; no need to make the macro unless you really wanna

File: code/datums/beam.dm (hunk 2 above)
Comment (duncathan) [inline code/datums/beam.dm]: ah yea I missed the "translation vectors" thing, lol. makes perfect sense in context.
might be helpful to clarify that it's in pixels but not really worth the effort if you won't be changing anything else

File: code/datums/beam.dm (hunk 3 above)
Comment (duncathan) [inline code/datums/beam.dm]: a better look at this line really just reveals to me that I honest to god have no idea why this equation is used or what it does but I'm sure that's just my own lack of knowledge with this stuff

Comment (RemieRichards) [inline code/datums/beam.dm]: well, this is two sin() calls, related to an angle along X
the two cos() calls below relate to an angle along Y.

I used to have a mnemonic for this, something like "cos I'm walking up the stairs (Y axis going up)" but it's been 2 years since I needed that mnemonic so it's gone lol.

tl;dr it's just trigonometry.

**Comment by duncathan** [comment]
:+1:

**Comment by Incoming5643** [comment]
While you don't have to, if you could find a solution to #12850 and add it to this it'd simplify things.

Target.loc is never actually used, so technically you only have to account for uses of origin.loc. The upside of tracking if target is a turf is that if you can be sure both ends of a beam are turfs you can assume they never need to be redrawn.

**Comment by MrStonedOne** [comment]
Please make a changelog,

You can add it to the PR's body rather then as a file now, see here: https://tgstation13.org/phpBB/viewtopic.php?f=5&t=5157

**Comment by duncathan** [comment]
@RemieRichards conflicts and changelog, please

**Comment by Razharas** [comment]
>Last activity 21 days ago

10 days left @RemieRichards 

tick

tock

tick

tock

**Comment by Zilenan91** [comment]
ded pr

**Comment by Shadowlight213** [comment]
@Zilenan91 2 days left till ded pr though.

**Comment by Zilenan91** [comment]
juan daey
